# Udacity Project - Bikesharing Dataset

## Data Wrangling and Data Visualisation

This data set includes information about individual rides made in a bike-sharing system covering the greater Boston area. 

Table of Contents
- Data Wrangling
- Explonatory Data Analysis
- Explanatory Data Analysis

# Data Wrangling


In [2]:
# import libraries 
import numpy as np
import pandas as pandas
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

**Structure of the dataset**

I have 12 csv's, for the 12 months of 2020. Ideally, I would like to see all 12 in one dataset, so it's easier to do explanatory analysis. 
Most of the months have 15 columns with a few having only 14. Need to make sure I'm not missing data so will reduce the ones with 15 rows to 14 before I merge them together. When I merge them, I will do it by quarters to help explore the data better. Different months have different amount of entries, but that's normal and I will not manipulate that.

I will look at the datasets statistics, datatypes, missing and duplicated values as well as visually assess the datasets to see if there is any need for cleaning.


**Insights to explore**

- How long does the average trip takes?
- Which are the most used start and end stations?
- When are the most trips taken in terms of time of day, day of the week, or month of the year?
- Does these depend on if a user is a subscriber or customer?

**Features in the dataset the will help support the investigation?**

Trip duration, Start time, End time, Start station, End Station, Usertype


## Assessment

In [9]:
# load in each month's dataset into a pandas dataframe
df_01 = pandas.read_csv('202001-bluebikes-tripdata.csv')
df_01.head()

tripduration                 starttime                  stoptime  \
0           478  2020-01-01 00:04:05.8090  2020-01-01 00:12:04.2370   
1           363  2020-01-01 00:04:45.6990  2020-01-01 00:10:49.0400   
2           284  2020-01-01 00:06:07.0630  2020-01-01 00:10:51.9240   
3           193  2020-01-01 00:06:13.8550  2020-01-01 00:09:27.8320   
4           428  2020-01-01 00:07:25.2950  2020-01-01 00:14:33.7800   

   start station id                           start station name  \
0               366                              Broadway T Stop   
1               219                  Boston East - 126 Border St   
2               219                  Boston East - 126 Border St   
3               396                         Main St at Beacon St   
4                60  Charles Circle - Charles St at Cambridge St   

   start station latitude  start station longitude  end station id  \
0               42.342781               -71.057473              93   
1               42.373312               -71.041020             212   
2               42.373312               -71.041020             212   
3               42.409330               -71.063819             387   
4               42.360793               -71.071190              49   

               end station name  end station latitude  end station longitude  \
0              JFK/UMass T Stop             42.320340             -71.051180   
1  Maverick Square - Lewis Mall             42.368844             -71.039778   
2  Maverick Square - Lewis Mall             42.368844             -71.039778   
3        Norman St at Kelvin St             42.409859             -71.066319   
4       Stuart St at Charles St             42.351146             -71.066289   

   bikeid    usertype  birth year  gender  
0    6005    Customer        1969       0  
1    3168  Subscriber        2000       1  
2    3985  Subscriber        2001       1  
3    2692  Subscriber        1978       1  
4    4978  Subscriber        1987       1

In [10]:
# assess shape
df_01.shape

(128598, 15)

In [11]:
# assess column names in each data set to see which column is missing
df_01.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128598 entries, 0 to 128597
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             128598 non-null  int64  
 1   starttime                128598 non-null  object 
 2   stoptime                 128598 non-null  object 
 3   start station id         128598 non-null  int64  
 4   start station name       128598 non-null  object 
 5   start station latitude   128598 non-null  float64
 6   start station longitude  128598 non-null  float64
 7   end station id           128598 non-null  int64  
 8   end station name         128598 non-null  object 
 9   end station latitude     128598 non-null  float64
 10  end station longitude    128598 non-null  float64
 11  bikeid                   128598 non-null  int64  
 12  usertype                 128598 non-null  object 
 13  birth year               128598 non-null  int64  
 14  gend

In [13]:
df_02 = pandas.read_csv('202002-bluebikes-tripdata.csv')
df_02.shape

(133235, 15)

In [16]:
# assess column names in each data set to see which column is missing
df_02.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133235 entries, 0 to 133234
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             133235 non-null  int64  
 1   starttime                133235 non-null  object 
 2   stoptime                 133235 non-null  object 
 3   start station id         133235 non-null  int64  
 4   start station name       133235 non-null  object 
 5   start station latitude   133235 non-null  float64
 6   start station longitude  133235 non-null  float64
 7   end station id           133235 non-null  int64  
 8   end station name         133235 non-null  object 
 9   end station latitude     133235 non-null  float64
 10  end station longitude    133235 non-null  float64
 11  bikeid                   133235 non-null  int64  
 12  usertype                 133235 non-null  object 
 13  birth year               133235 non-null  int64  
 14  gend

In [18]:
df_03 = pandas.read_csv('202003-bluebikes-tripdata.csv')
df_03.shape

(107350, 15)

In [19]:
# assess column names in each data set to see which column is missing
df_03.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107350 entries, 0 to 107349
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             107350 non-null  int64  
 1   starttime                107350 non-null  object 
 2   stoptime                 107350 non-null  object 
 3   start station id         107350 non-null  int64  
 4   start station name       107350 non-null  object 
 5   start station latitude   107350 non-null  float64
 6   start station longitude  107350 non-null  float64
 7   end station id           107350 non-null  int64  
 8   end station name         107350 non-null  object 
 9   end station latitude     107350 non-null  float64
 10  end station longitude    107350 non-null  float64
 11  bikeid                   107350 non-null  int64  
 12  usertype                 107350 non-null  object 
 13  birth year               107350 non-null  int64  
 14  gend

In [21]:
df_04 = pandas.read_csv('202004-bluebikes-tripdata.csv')
df_04.shape

(46793, 15)

In [22]:
# assess column names in each data set to see which column is missing
df_04.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46793 entries, 0 to 46792
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   tripduration             46793 non-null  int64  
 1   starttime                46793 non-null  object 
 2   stoptime                 46793 non-null  object 
 3   start station id         46793 non-null  int64  
 4   start station name       46793 non-null  object 
 5   start station latitude   46793 non-null  float64
 6   start station longitude  46793 non-null  float64
 7   end station id           46793 non-null  int64  
 8   end station name         46793 non-null  object 
 9   end station latitude     46793 non-null  float64
 10  end station longitude    46793 non-null  float64
 11  bikeid                   46793 non-null  int64  
 12  usertype                 46793 non-null  object 
 13  birth year               46793 non-null  int64  
 14  gender                

In [32]:
df_05 = pandas.read_csv('202005-bluebikes-tripdata.csv')
df_05.shape

(124879, 14)

In [33]:
# assess column names in each data set to see which column is missing
df_05.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124879 entries, 0 to 124878
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             124879 non-null  int64  
 1   starttime                124879 non-null  object 
 2   stoptime                 124879 non-null  object 
 3   start station id         124879 non-null  int64  
 4   start station name       124879 non-null  object 
 5   start station latitude   124879 non-null  float64
 6   start station longitude  124879 non-null  float64
 7   end station id           124879 non-null  int64  
 8   end station name         124879 non-null  object 
 9   end station latitude     124879 non-null  float64
 10  end station longitude    124879 non-null  float64
 11  bikeid                   124879 non-null  int64  
 12  usertype                 124879 non-null  object 
 13  postal code              110498 non-null  object 
dtypes: f

In [34]:
df_06 = pandas.read_csv('202006-bluebikes-tripdata.csv')
df_06.shape

(191843, 14)

In [35]:
# assess column names in each data set to see which column is missing
df_06.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 191843 entries, 0 to 191842
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             191843 non-null  int64  
 1   starttime                191843 non-null  object 
 2   stoptime                 191843 non-null  object 
 3   start station id         191843 non-null  int64  
 4   start station name       191843 non-null  object 
 5   start station latitude   191843 non-null  float64
 6   start station longitude  191843 non-null  float64
 7   end station id           191843 non-null  int64  
 8   end station name         191843 non-null  object 
 9   end station latitude     191843 non-null  float64
 10  end station longitude    191843 non-null  float64
 11  bikeid                   191843 non-null  int64  
 12  usertype                 191843 non-null  object 
 13  postal code              170883 non-null  object 
dtypes: f

In [36]:
df_07 = pandas.read_csv('202007-bluebikes-tripdata.csv')
df_07.shape

(259726, 14)

In [37]:
# assess column names in each data set to see which column is missing
df_07.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 259726 entries, 0 to 259725
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             259726 non-null  int64  
 1   starttime                259726 non-null  object 
 2   stoptime                 259726 non-null  object 
 3   start station id         259726 non-null  int64  
 4   start station name       259726 non-null  object 
 5   start station latitude   259726 non-null  float64
 6   start station longitude  259726 non-null  float64
 7   end station id           259726 non-null  int64  
 8   end station name         259726 non-null  object 
 9   end station latitude     259726 non-null  float64
 10  end station longitude    259726 non-null  float64
 11  bikeid                   259726 non-null  int64  
 12  usertype                 259726 non-null  object 
 13  postal code              230479 non-null  object 
dtypes: f

In [38]:
df_08 = pandas.read_csv('202008-bluebikes-tripdata.csv')
df_08.shape

(289033, 14)

In [146]:
# assess column names in each data set to see which column is missing
df_08.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 289033 entries, 0 to 289032
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             289033 non-null  int64  
 1   starttime                289033 non-null  object 
 2   stoptime                 289033 non-null  object 
 3   start station id         289033 non-null  int64  
 4   start station name       289033 non-null  object 
 5   start station latitude   289033 non-null  float64
 6   start station longitude  289033 non-null  float64
 7   end station id           289033 non-null  int64  
 8   end station name         289033 non-null  object 
 9   end station latitude     289033 non-null  float64
 10  end station longitude    289033 non-null  float64
 11  bikeid                   289033 non-null  int64  
 12  usertype                 289033 non-null  object 
 13  postal code              264273 non-null  object 
dtypes: f

In [39]:
df_09 = pandas.read_csv('202009-bluebikes-tripdata.csv')
df_09.shape

(307853, 14)

In [40]:
# assess column names in each data set to see which column is missing
df_09.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307853 entries, 0 to 307852
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             307853 non-null  int64  
 1   starttime                307853 non-null  object 
 2   stoptime                 307853 non-null  object 
 3   start station id         307853 non-null  int64  
 4   start station name       307853 non-null  object 
 5   start station latitude   307853 non-null  float64
 6   start station longitude  307853 non-null  float64
 7   end station id           307853 non-null  int64  
 8   end station name         307853 non-null  object 
 9   end station latitude     307853 non-null  float64
 10  end station longitude    307853 non-null  float64
 11  bikeid                   307853 non-null  int64  
 12  usertype                 307853 non-null  object 
 13  postal code              284701 non-null  object 
dtypes: f

In [42]:
df_10 = pandas.read_csv('202010-bluebikes-tripdata.csv')
df_10.shape

(248424, 14)

In [43]:
# assess column names in each data set to see which column is missing
df_10.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248424 entries, 0 to 248423
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             248424 non-null  int64  
 1   starttime                248424 non-null  object 
 2   stoptime                 248424 non-null  object 
 3   start station id         248424 non-null  int64  
 4   start station name       248424 non-null  object 
 5   start station latitude   248424 non-null  float64
 6   start station longitude  248424 non-null  float64
 7   end station id           248424 non-null  int64  
 8   end station name         248424 non-null  object 
 9   end station latitude     248424 non-null  float64
 10  end station longitude    248424 non-null  float64
 11  bikeid                   248424 non-null  int64  
 12  usertype                 248424 non-null  object 
 13  postal code              230070 non-null  object 
dtypes: f

In [45]:
df_11 = pandas.read_csv('202011-bluebikes-tripdata.csv')
df_11.shape

(161712, 14)

In [46]:
# assess column names in each data set to see which column is missing
df_11.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161712 entries, 0 to 161711
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             161712 non-null  int64  
 1   starttime                161712 non-null  object 
 2   stoptime                 161712 non-null  object 
 3   start station id         161712 non-null  int64  
 4   start station name       161712 non-null  object 
 5   start station latitude   161712 non-null  float64
 6   start station longitude  161712 non-null  float64
 7   end station id           161712 non-null  int64  
 8   end station name         161712 non-null  object 
 9   end station latitude     161712 non-null  float64
 10  end station longitude    161712 non-null  float64
 11  bikeid                   161712 non-null  int64  
 12  usertype                 161712 non-null  object 
 13  postal code              151233 non-null  object 
dtypes: f

In [47]:
df_12 = pandas.read_csv('202012-bluebikes-tripdata.csv')
df_12.shape

(74002, 14)

In [48]:
# assess column names in each data set to see which column is missing
df_12.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74002 entries, 0 to 74001
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   tripduration             74002 non-null  int64  
 1   starttime                74002 non-null  object 
 2   stoptime                 74002 non-null  object 
 3   start station id         74002 non-null  int64  
 4   start station name       74002 non-null  object 
 5   start station latitude   74002 non-null  float64
 6   start station longitude  74002 non-null  float64
 7   end station id           74002 non-null  int64  
 8   end station name         74002 non-null  object 
 9   end station latitude     74002 non-null  float64
 10  end station longitude    74002 non-null  float64
 11  bikeid                   74002 non-null  int64  
 12  usertype                 74002 non-null  object 
 13  postal code              71109 non-null  object 
dtypes: float64(4), int64(4

### Summary of Assessment

### Quality Issues
*All datasets*
- `starttime` and `stopttime` column is in object format, it should be datetime

### Tidiness Issues

- merge datasets (concat)
- remove birthyear, postcalcode and gender columns as it only has partial month's data coverage, not the full year

I have to state here, that I haven't gone through all the assessment like checking for missing values, etc. I would like to do that after merging the datasets together as it will be a lot easier like that so I assume that there will be a Cleaning Part 2 upon succeeding to merge the datasets together.



## Cleaning dataset

### Define 
- merge datasets (concat)

Action: find best method to merge datasets

### Code

In [49]:
# make a copy of all datasets to be sure I can revert if things go wrong
df_01_copy = df_01.copy()
df_02_copy = df_02.copy()
df_03_copy = df_03.copy()
df_04_copy = df_04.copy()
df_05_copy = df_05.copy()
df_06_copy = df_06.copy()
df_07_copy = df_07.copy()
df_08_copy = df_08.copy()
df_09_copy = df_09.copy()
df_10_copy = df_10.copy()
df_11_copy = df_11.copy()
df_12_copy = df_12.copy()

Check each copied datasets shape so I can see if I got all the data after merging

In [50]:
df_01_copy.shape

(128598, 15)

In [51]:
df_02_copy.shape

(133235, 15)

In [52]:
df_03_copy.shape

(107350, 15)

In [53]:
df_04_copy.shape

(46793, 15)

In [54]:
df_05_copy.shape

(124879, 14)

In [55]:
df_06_copy.shape

(191843, 14)

In [56]:
df_07_copy.shape

(259726, 14)

In [57]:
df_08_copy.shape

(289033, 14)

In [58]:
df_09_copy.shape

(307853, 14)

In [59]:
df_10_copy.shape

(248424, 14)

In [60]:
df_11_copy.shape

(161712, 14)

In [61]:
df_12_copy.shape

(74002, 14)

Check if copies have the right columns

In [62]:
df_01_copy.head()

tripduration                 starttime                  stoptime  \
0           478  2020-01-01 00:04:05.8090  2020-01-01 00:12:04.2370   
1           363  2020-01-01 00:04:45.6990  2020-01-01 00:10:49.0400   
2           284  2020-01-01 00:06:07.0630  2020-01-01 00:10:51.9240   
3           193  2020-01-01 00:06:13.8550  2020-01-01 00:09:27.8320   
4           428  2020-01-01 00:07:25.2950  2020-01-01 00:14:33.7800   

   start station id                           start station name  \
0               366                              Broadway T Stop   
1               219                  Boston East - 126 Border St   
2               219                  Boston East - 126 Border St   
3               396                         Main St at Beacon St   
4                60  Charles Circle - Charles St at Cambridge St   

   start station latitude  start station longitude  end station id  \
0               42.342781               -71.057473              93   
1               42.373312               -71.041020             212   
2               42.373312               -71.041020             212   
3               42.409330               -71.063819             387   
4               42.360793               -71.071190              49   

               end station name  end station latitude  end station longitude  \
0              JFK/UMass T Stop             42.320340             -71.051180   
1  Maverick Square - Lewis Mall             42.368844             -71.039778   
2  Maverick Square - Lewis Mall             42.368844             -71.039778   
3        Norman St at Kelvin St             42.409859             -71.066319   
4       Stuart St at Charles St             42.351146             -71.066289   

   bikeid    usertype  birth year  gender  
0    6005    Customer        1969       0  
1    3168  Subscriber        2000       1  
2    3985  Subscriber        2001       1  
3    2692  Subscriber        1978       1  
4    4978  Subscriber        1987       1

In [63]:
df_06_copy.head()

tripduration                 starttime                  stoptime  \
0          1160  2020-06-01 00:00:25.6240  2020-06-01 00:19:45.7010   
1          1419  2020-06-01 00:01:09.3800  2020-06-01 00:24:48.9520   
2          1093  2020-06-01 00:01:29.4790  2020-06-01 00:19:43.3530   
3          1067  2020-06-01 00:01:35.8540  2020-06-01 00:19:23.3250   
4          1391  2020-06-01 00:01:51.0390  2020-06-01 00:25:02.8460   

   start station id                   start station name  \
0               192              Purchase St at Pearl St   
1               355  Bennington St at Constitution Beach   
2               131                Jackson Square T Stop   
3               131                Jackson Square T Stop   
4               355  Bennington St at Constitution Beach   

   start station latitude  start station longitude  end station id  \
0               42.354659               -71.053181             366   
1               42.385224               -71.010631             355   
2               42.322931               -71.100141             331   
3               42.322931               -71.100141             331   
4               42.385224               -71.010631             355   

                      end station name  end station latitude  \
0                      Broadway T Stop             42.342781   
1  Bennington St at Constitution Beach             42.385224   
2           Huntington Ave at Mass Art             42.336586   
3           Huntington Ave at Mass Art             42.336586   
4  Bennington St at Constitution Beach             42.385224   

   end station longitude  bikeid    usertype postal code  
0             -71.057473    2831    Customer         NaN  
1             -71.010631    5696    Customer       02128  
2             -71.098870    3922  Subscriber       10570  
3             -71.098870    3361  Subscriber       10570  
4             -71.010631    3621    Customer       01902

In [64]:
# concat the data sets
df_concatenated_columns_all = pandas.concat([df_01_copy, df_02_copy, df_03_copy, df_04_copy, df_05_copy, df_06_copy, df_07_copy, df_08_copy, df_09_copy, df_10_copy, df_11_copy, df_12_copy,])


### Test

In [65]:
df_concatenated_columns_all.head()

tripduration                 starttime                  stoptime  \
0           478  2020-01-01 00:04:05.8090  2020-01-01 00:12:04.2370   
1           363  2020-01-01 00:04:45.6990  2020-01-01 00:10:49.0400   
2           284  2020-01-01 00:06:07.0630  2020-01-01 00:10:51.9240   
3           193  2020-01-01 00:06:13.8550  2020-01-01 00:09:27.8320   
4           428  2020-01-01 00:07:25.2950  2020-01-01 00:14:33.7800   

   start station id                           start station name  \
0               366                              Broadway T Stop   
1               219                  Boston East - 126 Border St   
2               219                  Boston East - 126 Border St   
3               396                         Main St at Beacon St   
4                60  Charles Circle - Charles St at Cambridge St   

   start station latitude  start station longitude  end station id  \
0               42.342781               -71.057473              93   
1               42.373312               -71.041020             212   
2               42.373312               -71.041020             212   
3               42.409330               -71.063819             387   
4               42.360793               -71.071190              49   

               end station name  end station latitude  end station longitude  \
0              JFK/UMass T Stop             42.320340             -71.051180   
1  Maverick Square - Lewis Mall             42.368844             -71.039778   
2  Maverick Square - Lewis Mall             42.368844             -71.039778   
3        Norman St at Kelvin St             42.409859             -71.066319   
4       Stuart St at Charles St             42.351146             -71.066289   

   bikeid    usertype  birth year  gender postal code  
0    6005    Customer      1969.0     0.0         NaN  
1    3168  Subscriber      2000.0     1.0         NaN  
2    3985  Subscriber      2001.0     1.0         NaN  
3    2692  Subscriber      1978.0     1.0         NaN  
4    4978  Subscriber      1987.0     1.0         NaN

In [66]:
df_concatenated_columns_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2073448 entries, 0 to 74001
Data columns (total 16 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   tripduration             int64  
 1   starttime                object 
 2   stoptime                 object 
 3   start station id         int64  
 4   start station name       object 
 5   start station latitude   float64
 6   start station longitude  float64
 7   end station id           int64  
 8   end station name         object 
 9   end station latitude     float64
 10  end station longitude    float64
 11  bikeid                   int64  
 12  usertype                 object 
 13  birth year               float64
 14  gender                   float64
 15  postal code              object 
dtypes: float64(6), int64(4), object(6)
memory usage: 268.9+ MB


In [67]:
df_concatenated_columns_all.shape

(2073448, 16)

In [68]:
df_all = df_concatenated_columns_all.copy()

In [69]:
df_all.shape


(2073448, 16)

### Define
- remove birth year, gender and postal code columns

Action: 
drop these columns permanently

### Code

In [70]:
df_all.drop(['birth year', 'postal code', 'gender'], axis=1, inplace=True)

### Test

In [71]:
df_all.head()

tripduration                 starttime                  stoptime  \
0           478  2020-01-01 00:04:05.8090  2020-01-01 00:12:04.2370   
1           363  2020-01-01 00:04:45.6990  2020-01-01 00:10:49.0400   
2           284  2020-01-01 00:06:07.0630  2020-01-01 00:10:51.9240   
3           193  2020-01-01 00:06:13.8550  2020-01-01 00:09:27.8320   
4           428  2020-01-01 00:07:25.2950  2020-01-01 00:14:33.7800   

   start station id                           start station name  \
0               366                              Broadway T Stop   
1               219                  Boston East - 126 Border St   
2               219                  Boston East - 126 Border St   
3               396                         Main St at Beacon St   
4                60  Charles Circle - Charles St at Cambridge St   

   start station latitude  start station longitude  end station id  \
0               42.342781               -71.057473              93   
1               42.373312               -71.041020             212   
2               42.373312               -71.041020             212   
3               42.409330               -71.063819             387   
4               42.360793               -71.071190              49   

               end station name  end station latitude  end station longitude  \
0              JFK/UMass T Stop             42.320340             -71.051180   
1  Maverick Square - Lewis Mall             42.368844             -71.039778   
2  Maverick Square - Lewis Mall             42.368844             -71.039778   
3        Norman St at Kelvin St             42.409859             -71.066319   
4       Stuart St at Charles St             42.351146             -71.066289   

   bikeid    usertype  
0    6005    Customer  
1    3168  Subscriber  
2    3985  Subscriber  
3    2692  Subscriber  
4    4978  Subscriber

### Define
- `starttime` and `stopttime` column is in object format, it should be datetime

Action: convert these to datetime

### Code

In [74]:
# convert starttime field into datetime
df_all['starttime'] = pandas.to_datetime(df_all['starttime'])

In [75]:
# convert stoptime field into datetime
df_all['stoptime'] = pandas.to_datetime(df_all['stoptime'])

### Test

In [76]:
df_all.dtypes

tripduration                        int64
starttime                  datetime64[ns]
stoptime                   datetime64[ns]
start station id                    int64
start station name                 object
start station latitude            float64
start station longitude           float64
end station id                      int64
end station name                   object
end station latitude              float64
end station longitude             float64
bikeid                              int64
usertype                           object
dtype: object

## Cleaning part 2

Let's first check if there are any outliers in the data.

In [77]:
# assess df_all statistics
df_all.describe()

tripduration  start station id  start station latitude  \
count  2.073448e+06      2.073448e+06            2.073448e+06   
mean   1.831438e+03      1.634481e+02            4.235597e+01   
std    2.425891e+04      1.357767e+02            3.452189e-02   
min    6.100000e+01      1.000000e+00            0.000000e+00   
25%    4.750000e+02      5.500000e+01            4.234522e+01   
50%    8.280000e+02      1.100000e+02            4.235560e+01   
75%    1.443000e+03      2.390000e+02            4.236567e+01   
max    3.879352e+06      4.990000e+02            4.241608e+01   

       start station longitude  end station id  end station latitude  \
count             2.073448e+06    2.073448e+06          2.073448e+06   
mean             -7.108868e+01    1.623836e+02          4.235589e+01   
std               5.628081e-02    1.358793e+02          3.455072e-02   
min              -7.122627e+01    1.000000e+00          0.000000e+00   
25%              -7.110594e+01    5.500000e+01          4.234522e+01   
50%              -7.108981e+01    1.080000e+02          4.235560e+01   
75%              -7.106959e+01    2.390000e+02          4.236567e+01   
max               0.000000e+00    4.990000e+02          4.241608e+01   

       end station longitude        bikeid  
count           2.073448e+06  2.073448e+06  
mean           -7.108840e+01  4.263332e+03  
std             5.626398e-02  1.272965e+03  
min            -7.122627e+01  3.100000e+01  
25%            -7.110567e+01  3.175000e+03  
50%            -7.108822e+01  4.256000e+03  
75%            -7.106894e+01  5.406000e+03  
max             0.000000e+00  6.724000e+03

There must be some zero values in the start latitude and end latitude columns, but since I'm not going to explore the data in those columns, I will not clean the data there.

In [78]:
# assess df_all's duplicated values
sum(df_all.duplicated())

0

In [79]:
# assess df_all's null values
df_all.isnull().sum()

tripduration               0
starttime                  0
stoptime                   0
start station id           0
start station name         0
start station latitude     0
start station longitude    0
end station id             0
end station name           0
end station latitude       0
end station longitude      0
bikeid                     0
usertype                   0
dtype: int64

It seems like the data set is clean enough to start my exploration towards the initial questions, but to be able to draw conclusions from the starttime column, I would like to extract a couple of values and create new columns.

Q: When are the most trips taken in terms of time of day, day of the week, or month of the year?
 
 ### Tidiness Issues

There is no way to see
- which day of the week the data was recorded
- if the day is a weekday or a weekend
- which months of the year the data was recorded
- which time of the day the data was recorded



### Define 
There is no way to see which day of the week the data was recorded

Action:
- Create a column for the days of the week 
- Convert values to date time and then to week day names



### Code

In [197]:
# create a column for the days of the week using the starttime column

df_all['weekday'] = df_all['starttime'].dt.dayofweek   

# convert values to date time and then to week day names

df_all['weekday'] = pandas.to_datetime(df_all['weekday'])
df_all['weekday'] = df_all['weekday'].dt.day_name()


### Test

In [196]:
df_all.head(5)

tripduration               starttime                stoptime  \
0           478 2020-01-01 00:04:05.809 2020-01-01 00:12:04.237   
1           363 2020-01-01 00:04:45.699 2020-01-01 00:10:49.040   
2           284 2020-01-01 00:06:07.063 2020-01-01 00:10:51.924   
3           193 2020-01-01 00:06:13.855 2020-01-01 00:09:27.832   
4           428 2020-01-01 00:07:25.295 2020-01-01 00:14:33.780   

   start station id                           start station name  \
0               366                              Broadway T Stop   
1               219                  Boston East - 126 Border St   
2               219                  Boston East - 126 Border St   
3               396                         Main St at Beacon St   
4                60  Charles Circle - Charles St at Cambridge St   

   start station latitude  start station longitude  end station id  \
0               42.342781               -71.057473              93   
1               42.373312               -71.041020             212   
2               42.373312               -71.041020             212   
3               42.409330               -71.063819             387   
4               42.360793               -71.071190              49   

               end station name  end station latitude  end station longitude  \
0              JFK/UMass T Stop             42.320340             -71.051180   
1  Maverick Square - Lewis Mall             42.368844             -71.039778   
2  Maverick Square - Lewis Mall             42.368844             -71.039778   
3        Norman St at Kelvin St             42.409859             -71.066319   
4       Stuart St at Charles St             42.351146             -71.066289   

   bikeid    usertype   weekday    month timeoftheday     type  
0    6005    Customer  Thursday  January        00:04  Weekday  
1    3168  Subscriber  Thursday  January        00:04  Weekday  
2    3985  Subscriber  Thursday  January        00:06  Weekday  
3    2692  Subscriber  Thursday  January        00:06  Weekday  
4    4978  Subscriber  Thursday  January        00:07  Weekday

### Define 
There is no way to see which day of the month the data was recorded

Action:
- Create a column for the months of the year
- Convert values to date time and then to month names

### Code

In [199]:
# create a column for the months of the year using the starttime column
df_all['month'] = pandas.DatetimeIndex(df_all['starttime']).month

# convert values to date time and then month names

df_all['month'] = pandas.to_datetime(df_all['month'], format='%m').dt.month_name()


### Test

In [187]:
df_all.head()

tripduration               starttime                stoptime  \
0           478 2020-01-01 00:04:05.809 2020-01-01 00:12:04.237   
1           363 2020-01-01 00:04:45.699 2020-01-01 00:10:49.040   
2           284 2020-01-01 00:06:07.063 2020-01-01 00:10:51.924   
3           193 2020-01-01 00:06:13.855 2020-01-01 00:09:27.832   
4           428 2020-01-01 00:07:25.295 2020-01-01 00:14:33.780   

   start station id                           start station name  \
0               366                              Broadway T Stop   
1               219                  Boston East - 126 Border St   
2               219                  Boston East - 126 Border St   
3               396                         Main St at Beacon St   
4                60  Charles Circle - Charles St at Cambridge St   

   start station latitude  start station longitude  end station id  \
0               42.342781               -71.057473              93   
1               42.373312               -71.041020             212   
2               42.373312               -71.041020             212   
3               42.409330               -71.063819             387   
4               42.360793               -71.071190              49   

               end station name  end station latitude  end station longitude  \
0              JFK/UMass T Stop             42.320340             -71.051180   
1  Maverick Square - Lewis Mall             42.368844             -71.039778   
2  Maverick Square - Lewis Mall             42.368844             -71.039778   
3        Norman St at Kelvin St             42.409859             -71.066319   
4       Stuart St at Charles St             42.351146             -71.066289   

   bikeid    usertype   weekday    month timeoftheday     type  
0    6005    Customer  Thursday  January        00:04  Weekday  
1    3168  Subscriber  Thursday  January        00:04  Weekday  
2    3985  Subscriber  Thursday  January        00:06  Weekday  
3    2692  Subscriber  Thursday  January        00:06  Weekday  
4    4978  Subscriber  Thursday  January        00:07  Weekday

### Define 
There is no way to see which time of the day the data was recorded

Action:
- Create a column for the time of the day


### Code

In [188]:
# create a column for the time of the day using the starttime column
df_all['timeoftheday']= df_all['starttime'].dt.strftime('%H:%M')

### Test

In [189]:
df_all.head()

tripduration               starttime                stoptime  \
0           478 2020-01-01 00:04:05.809 2020-01-01 00:12:04.237   
1           363 2020-01-01 00:04:45.699 2020-01-01 00:10:49.040   
2           284 2020-01-01 00:06:07.063 2020-01-01 00:10:51.924   
3           193 2020-01-01 00:06:13.855 2020-01-01 00:09:27.832   
4           428 2020-01-01 00:07:25.295 2020-01-01 00:14:33.780   

   start station id                           start station name  \
0               366                              Broadway T Stop   
1               219                  Boston East - 126 Border St   
2               219                  Boston East - 126 Border St   
3               396                         Main St at Beacon St   
4                60  Charles Circle - Charles St at Cambridge St   

   start station latitude  start station longitude  end station id  \
0               42.342781               -71.057473              93   
1               42.373312               -71.041020             212   
2               42.373312               -71.041020             212   
3               42.409330               -71.063819             387   
4               42.360793               -71.071190              49   

               end station name  end station latitude  end station longitude  \
0              JFK/UMass T Stop             42.320340             -71.051180   
1  Maverick Square - Lewis Mall             42.368844             -71.039778   
2  Maverick Square - Lewis Mall             42.368844             -71.039778   
3        Norman St at Kelvin St             42.409859             -71.066319   
4       Stuart St at Charles St             42.351146             -71.066289   

   bikeid    usertype   weekday    month timeoftheday     type  
0    6005    Customer  Thursday  January        00:04  Weekday  
1    3168  Subscriber  Thursday  January        00:04  Weekday  
2    3985  Subscriber  Thursday  January        00:06  Weekday  
3    2692  Subscriber  Thursday  January        00:06  Weekday  
4    4978  Subscriber  Thursday  January        00:07  Weekday

### Define 
There is no way to see if the day of the week is a weekday or a weekend

Action:
- Create a column to state if it's a weekday or weekend

### Code

In [190]:
df_all['type'] = df_all.starttime.dt.weekday.apply(lambda x: 'Weekday' if x < 5 else 'Weekend')

### Test

In [194]:
df_all.head(1500)


tripduration               starttime                stoptime  \
0              478 2020-01-01 00:04:05.809 2020-01-01 00:12:04.237   
1              363 2020-01-01 00:04:45.699 2020-01-01 00:10:49.040   
2              284 2020-01-01 00:06:07.063 2020-01-01 00:10:51.924   
3              193 2020-01-01 00:06:13.855 2020-01-01 00:09:27.832   
4              428 2020-01-01 00:07:25.295 2020-01-01 00:14:33.780   
...            ...                     ...                     ...   
1495          1230 2020-01-01 21:51:42.091 2020-01-01 22:12:12.097   
1496           903 2020-01-01 21:55:38.713 2020-01-01 22:10:41.897   
1497           988 2020-01-01 21:57:33.631 2020-01-01 22:14:02.276   
1498           698 2020-01-01 21:59:39.092 2020-01-01 22:11:17.477   
1499           279 2020-01-01 21:59:44.437 2020-01-01 22:04:24.306   

      start station id                                 start station name  \
0                  366                                    Broadway T Stop   
1                  219                        Boston East - 126 Border St   
2                  219                        Boston East - 126 Border St   
3                  396                               Main St at Beacon St   
4                   60        Charles Circle - Charles St at Cambridge St   
...                ...                                                ...   
1495                30      Brigham Circle - Francis St at Huntington Ave   
1496               190                  Nashua Street at Red Auerbach Way   
1497                 6                             Cambridge St at Joy St   
1498               104  Harvard University Radcliffe Quadrangle at She...   
1499                74                Harvard Square at Mass Ave/ Dunster   

      start station latitude  start station longitude  end station id  \
0                  42.342781               -71.057473              93   
1                  42.373312               -71.041020             212   
2                  42.373312               -71.041020             212   
3                  42.409330               -71.063819             387   
4                  42.360793               -71.071190              49   
...                      ...                      ...             ...   
1495               42.334629               -71.104079              90   
1496               42.365673               -71.064263              78   
1497               42.361212               -71.065306             342   
1498               42.380287               -71.125107             221   
1499               42.373268               -71.118579             180   

                                 end station name  end station latitude  \
0                                JFK/UMass T Stop             42.320340   
1                    Maverick Square - Lewis Mall             42.368844   
2                    Maverick Square - Lewis Mall             42.368844   
3                          Norman St at Kelvin St             42.409859   
4                         Stuart St at Charles St             42.351146   
...                                           ...                   ...   
1495  Lechmere Station at Cambridge St / First St             42.370677   
1496                    Union Square - Somerville             42.379671   
1497                     Boylston St at Jersey St             42.344651   
1498        Verizon Innovation Hub 10 Ware Street             42.372509   
1499                                    Mt Auburn             42.374786   

      end station longitude  bikeid    usertype   weekday    month  \
0                -71.051180    6005    Customer  Thursday  January   
1                -71.039778    3168  Subscriber  Thursday  January   
2                -71.039778    3985  Subscriber  Thursday  January   
3                -71.066319    2692  Subscriber  Thursday  January   
4                -71.066289    4978  Subscriber  Thursday  January   
...                     ...     ...       

## Explore Univariate Variables

In this section, I'll investigate distributions of individual variables. My aim is to check if there are any unusual points or outliers, take a deeper look to clean things up and prepare myself to look at relationships between variables.

I will look into the following:
- How long does the average trip takes?
- Which are the most used start and end stations?
- When are the most trips taken in terms of time of day, day of the week, or month of the year?
- Does these depend on if a user is a subscriber or customer?


In [102]:
# asses how long does the average trip take
df_all['tripduration'].mean()

1831.438460477427

The average trip takes 1831 seconds, which is approx. 30 minutes.

In [97]:
# assess trip duration column's 15 most frequent values in descending order
df_all['tripduration'].value_counts().sort_values(ascending=False)[:15]


397    1897
382    1884
410    1879
399    1878
366    1873
376    1871
361    1866
391    1865
393    1863
380    1857
360    1857
352    1853
405    1851
368    1850
346    1847
Name: tripduration, dtype: int64

The most frequent trip duration shows values between 1847 and 1897 seconds which is around 30 minutes as well.

In [201]:
# assess start station id column's 10 most frequent values in descending order
df_all['start station id'].value_counts().sort_values(ascending=False)[:10]


68     34220
60     33094
67     29301
46     27899
107    23308
342    22866
178    22503
47     22194
74     21111
53     20713
Name: start station id, dtype: int64

In [227]:
df_all.loc[df_all['start station id'] == 34220, ['start station name']]

Empty DataFrame
Columns: [start station name]
Index: []

The most frequently used start stations are:

In [99]:
# assess end station id column's 15 most frequent values in descending order
df_all['end station id'].value_counts().sort_values(ascending=False)[:15]


68     35158
60     33450
46     28040
67     27574
107    27111
342    22970
74     22725
47     21995
190    21604
11     21473
178    21351
53     20595
41     19466
58     19447
179    18923
Name: end station id, dtype: int64

The most frequently used end stations are:

In [82]:
df_all.

SyntaxError: invalid syntax (<ipython-input-82-3d28575df74a>, line 1)